### PR TITLE
Add migration lifecycle hooks

### PR DESCRIPTION
**Pull request by timorthi**
This is an attempt at addressing the use cases described in #2983 . 

This adds the ability for the user to define a `beforeAll`, `afterAll`, `beforeEach`, `afterEach` hook during migrations.

### CODE REVIEW

File: lib/migrations/migrate/Migrator.js [hunk 1]
```diff
@@ -583,12 +602,9 @@ function getNewMigrations(migrationSource, all, completed) {
   });
 }
 
-function checkPromise(logger, migrationPromise, name, commitFn) {
+function checkPromise(logger, migrationPromise, name) {
   if (!migrationPromise || typeof migrationPromise.then !== 'function') {
     logger.warn(`migration ${name} did not return a promise`);
-    if (commitFn) {
-      commitFn();
-    }
```
Comment (timorthi) [inline lib/migrations/migrate/Migrator.js:591]: I removed the explicit commit so that we can run the `afterEach` hook in the same transaction (line 527)

File: test/integration2/migrate/migration-integration.spec.js [hunk 2]
```diff
@@ -527,7 +528,131 @@ describe('Migrations', function () {
           });
         });
 
+        describe('knex.migrate.latest - lifecycle hooks', function () {
```
Comment (rluvaton) [inline test/integration2/migrate/migration-integration.spec.js:531]: Could you please add tests that:
1. if one of the hooks fails the migration is rolled back
2. If one of the hook fails the next hooks/migrations are not executed:
    1. if `beforeAll` failed, the `beforeEach`, the migration itself, `afterEach` and `afterAll` are not called
    2. if `beforeEach` failed, the migration itself, `afterEach` and `afterAll` are not called
    3. if the migration itself failed, the `afterEach` and `afterAll` are not called
    4. if the `afterEach` failed, `afterAll` is not called
3. test with all the hooks and check execution order - `beforeAll` -> `beforeEach` -> `afterEach` -> `afterAll`
4. no hooks are being called if no pending migrations

Comment (rluvaton) [inline test/integration2/migrate/migration-integration.spec.js:531]: I think this should be in a separate file for life cycle hooks

File: test/integration2/migrate/migration-integration.spec.js [hunk 3]
```diff
@@ -527,7 +528,131 @@ describe('Migrations', function () {
...
+            });
+          });
+        });
+
         describe('knex.migrate.rollback', function () {
+          before(() => {
+            return knex.migrate.latest({
+              directory: 'test/integration2/migrate/test',
+            });
+          });
```
Comment (rluvaton) [inline test/integration2/migrate/migration-integration.spec.js:654]: Why is this needed all of a sudden?

**Review by rluvaton — COMMENTED** [review]
Looking good! thanks for the great job!
could you please also update the types?

left few comments and I also think docs should be updated as well (Cc @kibertoad)


Also few questions:
1. If you have 2 migrations, and the second one failed, should you run the `afterAll` should you rollback the `beforeAll`?
2. In case transaction are disabled for migrations, should we allow hooks?

File: test/integration2/migrate/migration-lifecycle-integration.spec.js [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+              'afterEach-20131019235242_migration_1.js',
+              'beforeEach-20131019235306_migration_2.js',
+              'afterEach-20131019235306_migration_2.js',
+              'afterAll',
+            ]);
+          });
+        });
+
+        // NOTE (PR #5541) This is disabled pending further discussion
+        describe.skip('when there are no pending migrations', function () {
```
Comment (timorthi) [inline test/integration2/migrate/migration-lifecycle-integration.spec.js:287]: This test is currently failing because there appears to be no short-circuiting check on the length of new migrations returned by `getNewMigrations`. https://github.com/knex/knex/blob/e2bc612179a8971d167bbb0d571f51fe4e602eac/lib/migrations/migrate/Migrator.js#L74

Is there a specific reason we don't short-circuit, or was this an oversight?

File: test/integration2/migrate/migration-integration.spec.js (hunk 2 above)
Comment (timorthi) [reply test/integration2/migrate/migration-integration.spec.js]: Great suggestions, thank you!

**Comment by timorthi** [comment]
@rluvaton 


> 1. If you have 2 migrations, and the second one failed, should you run the afterAll should you rollback the beforeAll?

In this scenario, I feel like it may be appropriate to abandon the afterAll and roll back the beforeAll. 

> 2. In case transaction are disabled for migrations, should we allow hooks?

I can't personally think of a situation in which someone would want to run these hooks without also running their migrations in a transaction, though I wouldn't be surprised if that's a use case for someone out there. Maybe for the initial release, we can require transactions to run hooks, and see if anyone raises an issue with that.

File: test/integration2/migrate/migration-lifecycle-integration.spec.js (hunk 4 above)
Comment (rluvaton) [reply test/integration2/migrate/migration-lifecycle-integration.spec.js:274]: I think without the hooks it did not matter, but now it does... you should add short circus

Comment (timorthi) [reply test/integration2/migrate/migration-lifecycle-integration.spec.js:274]: Done!

Comment (timorthi) [reply test/integration2/migrate/migration-lifecycle-integration.spec.js]: Done! I've also relocated the `beforeAll`/`afterAll` hook calls to a place that's more DRY.

File: test/integration2/migrate/migration-lifecycle-integration.spec.js [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+
+            await knex.migrate
+              .latest({
+                directory: 'test/integration2/migrate/test',
+                beforeEach,
+                afterEach,
+                afterAll,
+              })
+              .catch((error) => {
+                expect(error.message).to.equal('force beforeEach hook failure');
```
Comment (rluvaton) [inline test/integration2/migrate/migration-lifecycle-integration.spec.js:158]: If the function won't reject this won't be called, saved the error in a value as well as if rejected or not

File: test/integration2/migrate/migration-lifecycle-integration.spec.js [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+                hasSecondTestTable: false,
+              },
+              {
+                hasFirstTestTable: true,
+                hasSecondTestTable: false,
+              },
+            ]);
+          });
+
+          it('does not run the migration or the afterEach/afterAll hooks if the hook fails', async function () {
```
Comment (rluvaton) [inline test/integration2/migrate/migration-lifecycle-integration.spec.js:143]:
```suggestion
          it('does not run the migration and the afterEach/afterAll hooks if the hook fails', async function () {
```

File: test/integration2/migrate/migration-lifecycle-integration.spec.js [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+const { getAllDbs, getKnexForDb } = require('../util/knex-instance-provider');
+
+describe('Migrations Lifecycle Hooks', function () {
+  getAllDbs().forEach((db) => {
+    describe(db, () => {
+      let knex;
+
+      // Force clean slate before each test
+      beforeEach(async () => {
+        rimraf.sync(path.join(__dirname, './migration'));
```
Comment (rluvaton) [inline test/integration2/migrate/migration-lifecycle-integration.spec.js:19]: Why sync?

**Review by rluvaton — COMMENTED** [review]
Could you please add test when transaction are disabled for migration?

File: lib/migrations/migrate/Migrator.js [hunk 8]
```diff
@@ -496,30 +502,40 @@ class Migrator {
...
           ) {
             this.knex.enableProcessing();
-            return this._transaction(
-              this.knex,
-              migrationContent,
-              direction,
-              name
-            );
+            return await this.knex.transaction(async (trx) => {
+              await beforeEach(trx, [migration]);
```
Comment (timorthi) [inline lib/migrations/migrate/Migrator.js:519]: The migration that's passed into the hook is wrapped in an array here to maintain consistent types between the `each` and `all` hooks (instead of passing in an array in `all` but a migration object in `each`)

File: test/integration2/migrate/migration-lifecycle-integration.spec.js (hunk 7 above)
Comment (timorthi) [reply test/integration2/migrate/migration-lifecycle-integration.spec.js:19]: Not 100% sure - this setup block was copied from `migration-integration.spec.js` and I chose not to change anything.

**Comment by timorthi** [comment]
@rluvaton 

> Could you please add test when transaction are disabled for migration?

I took a closer look at the code and it appears that the user is allowed to customize whether each migration is run in a transaction or not: https://github.com/knex/knex/blob/f7a31477e203dd9b00c36b841de204c5d109f6c1/lib/migrations/migrate/Migrator.js#L480-L490

Because of this, I think it may be simpler to just run the lifecycle hooks all the time, even if transactions are disabled. Otherwise, we will be dealing with a bunch of edge cases regarding when or when not to run lifecycle hooks. What do you think and would you still like me to test the non-transaction case if we choose to always run hooks?

**Comment by rluvaton** [comment]
Can you please update the types

**Comment by timorthi** [comment]
@rluvaton I've updated `index.d.ts`, did I miss any other type definitions?

For CI, it looks like this particular step (Legacy Types Linting) has been failing for the last 3 weeks, unrelated to this PR https://github.com/knex/knex/actions/workflows/linting-legacy.yml

**Review by rluvaton — APPROVED** [review]
Thank you for your contribution

**Comment by michaelwiles** [comment]
Please can this be merged? I need this as well.

**Comment by m33ch** [comment]
Hello and thank you all for your contribution! I really needed this feature.

I noticed that the change has been merged into master but I see that a new version hasn't been pubilshed to npm.
Approximately, when is a new release on npm expected?

Thanks a lot!
